naga/src/back/msl/writer.rs

```diff
@@ -1497,6 +1497,63 @@ impl<W: Write> Writer<W> {
         Ok(())
     }
 
+    /// Emit code for the WGSL functions `pack4x{I, U}8[Clamp]`.
+    fn put_pack4x8(
+        &mut self,
+        arg: Handle<crate::Expression>,
+        context: &ExpressionContext<'_>,
+        was_signed: bool,
+        clamp_bounds: Option<(&str, &str)>,
+    ) -> Result<(), Error> {
+        if context.lang_version >= (2, 1) {
+            let packed_type = if was_signed {
+                "packed_char4"
+            } else {
+                "packed_uchar4"
+            };
+            // Metal uses little endian byte order, which matches what WGSL expects here.
+            write!(self.out, "as_type<uint>({packed_type}(")?;
+            if let Some((min, max)) = clamp_bounds {
+                // Clamping a vector to scalar bounds works and operates component-wise.
+                write!(self.out, "{NAMESPACE}::clamp(")?;
+                self.put_expression(arg, context, true)?;
+                write!(self.out, ", {min}, {max})")?;
+            } else {
+                self.put_expression(arg, context, true)?;
+            }
```

Comment:
Can we move `write_arg` up and call it here as well?

Reply:
You're right, this was silly.